the-collab-lab/tcl-6-smart-shopping-list
# Pja sb sort by next purchase.

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Related Issue



closes #35

## Description



Shopping list items are displayed in order of how soon the user is likely to need to buy each of them again.

The list is sorted by default.

Item are grouped into these categories:

- Soon (next purchase estimated in fewer than 8 days)
- Kind of soon (next purchase estimated in 8 to 14 days)
- Not soon (next purchase estimated in 15 or more days)
- Inactive (the item was last purchased longer ago than twice the next estimate)

Each category is visually distinguished using different text colors, fonts, and/or font weights. Colorblind accessibility was checked using the [Chrome Colorblinding](https://chrome.google.com/webstore/detail/colorblinding/dgbgleaofjainknadoffbjkclicbbgaa) extension. Monochromacy seems to be the hardest to read, but the differentiation in font styling helps.

VoiceOver (VO) on Mac was used to test screen readability. VO will read the item name and the number of days until the next purchase. The next purchase estimate is hidden from sighted users, but VO can read it. By default VO reads where in the list the items falls, e.g. "Frog. Next purchase in 5 days. Item 4 of 16".

## Steps taken

- [x] Determine our sorting criteria
- [x] Sort items alphabetically
- [x] Sort items by days until next purchase
- [x] Sort inactive items
- [x] Add appropriate class names to list items
- [x] Visually distinguish  item groupings
- [x] Markup for screen readers

## Updates

### Before


![Screen Shot 2020-05-11 at 8 43 56 AM](https://user-images.githubusercontent.com/7929227/81581445-9b040a00-9363-11ea-9dee-554b759265ca.png)

### After

![Screen Shot 2020-05-15 at 3 13 37 PM](https://user-images.githubusercontent.com/7929227/82100546-bb62fa00-96be-11ea-8c8a-02b85829a762.png)


## Acceptance Criteria



1. Items in the list are shown visually grouped in the following order: Soon, Kind of soon, Not soon, Inactive
2. Within each grouping, items should be sorted by the estimated number of days until next purchase
3. Items with the same number of estimated days until next purchase should be sorted alphabetically
4. Items in the different states should be visually distinct
5. Items in the different states should be described distinctly when read by a screen reader

## Testing Steps / QA Criteria



View the deployed branch [here](https://deploy-preview-39--tcl-6-smart-shopping-list.netlify.app/).
To test with VoiceOver (VO):

- Make sure the Shopping app is opened in its own browser window. (It is a little less difficult to test VO this way.)
- Open VO with `cmd + F5` or in the accessibility panel in system preferences.
- Once VO is open press `ctrl + opt + U` to open VO Rotor.
- Use the left or right arrow keys to get to the Articles menu.
- Use the down arrow to select `list` in the menu.
- Hit enter.
- Use `ctrl + opt right arrow` to move through the list items and buttons.
- Use `ctrl + opt left arrow` to move back previous elements.
